### ISSUE TITLE
DataStore does not sync updated Models where keys are part of Composite keys

### ISSUE THREAD

**Issue by meducati** [aws-amplify/amplify-js#7714]
**Describe the bug**
Saving an updated model works locally, but sync to backend fails when the updated keys are also part of a composite key.

**To Reproduce**
1. Create schema model with @key and fields: ["alpha", "beta"]
2. Update an existing DataStore item with at least one of the field keys (Datastore.save, modelname.copyof...)
3. DataStore will update as expected
4. Sync will throw a warning (see below), due to the auto-generated resolver. The reason however, is that  the "outboxMutationEnqueued" element is not including all updated keys in the "data" variable

**Expected behavior**
Update should work

**Code Snippet**
Schema (part of):
type CADFeature
@model
@key(name: "FeaturesByFile", fields: ["fileId","tool","status"], queryField: "getFeaturesByFileId")
@auth(rules: [
    {allow: owner},
    {allow: groups, groups: ["admin"] },
    {allow: groups, groups: ["users"], operations: [read] },
    {allow: private, provider: iam, operations: [read] }
  ])
{
id: ID!
status: FeatureStatus!
characteristic: String
notes: String
tool: FeatureTool!
...

Example template ("status" and "tool" are fields in the "FeaturesByFile" @key):
```javascript
await DataStore.save(
    CADFeature.copyOf(features[0], updated => {
        updated.status = FeatureStatus.DELETED;
        updated.tool = features[0].tool;
        updated.characteristic = 'whatever';
        updated.notes = 'whatever2';
    })
);
```
**Does NOT work: (EDIT: These are both keys that are part of a composite key)**
1. updated.status = FeatureStatus.DELETED only
2. updated.tool = features[0].tool only
3. updated.status AND updated.tool

**Does work: (EDIT: These are both keys that are NOT part of a composite key)**
1. updated.characteristic only
2. updated.notes only
3. updated.characteristic AND updated.status

**Screenshots**
This DataStore warning originates in the Resolver which is correct. (In my previous build, this check was not generated, so must have come from in a recent aws-amplify version)). _Note that I updated BOTH status and tool for this example, but only status is submitted in the sync operation._
![DatastoreWarning](https://user-images.githubusercontent.com/63409162/107383821-a7cfad00-6af1-11eb-9a9e-c73abb34e8d0.JPG)

As seen here, "tool" is never included in the pending mutation (but as mentioned before, tool has been updated in the local DataStore).
![outBoxMutationEnqued](https://user-images.githubusercontent.com/63409162/107385559-490b3300-6af3-11eb-9054-e6be2862144b.JPG)

**What is Configured?**
aws cli 4.41.2 (EDIT: The generated resolvers might have com from a slightly older version, but younger than 4.21.0)

"dependencies": {
        "aws-amplify": "^3.3.17",
        "aws-sdk": "^2.736.0",
        "save-dev": "0.0.1-security"
    },
    "devDependencies": {
        "clean-webpack-plugin": "^3.0.0",
        "copy-webpack-plugin": "^6.1.1",
        "css-loader": "^3.5.3",
        "html-webpack-plugin": "^3.2.0",
        "ini": "^1.3.5",
        "inquirer": "^6.5.1",
        "style-loader": "^1.2.1",
        "terser-webpack-plugin": "^4.1.0",
        "webpack": "^4.44.1",
        "webpack-cli": "^3.3.12",
        "webpack-dev-server": "^3.11.0",
        "webpack-merge": "^5.1.1",
        "webpack-pwa-manifest": "^4.1.1",
        "workbox-webpack-plugin": "^5.1.3"
    },

**Comment by meducati** [comment]
So I have been digging into this issue for the last couple of days, trying to understand what is happening in the backend.

The culprit has to do with that the "patches" array in Datastore.js not picking up all my updated keys. I believe its actually just the cause **on the second key that is part of a composite key**:

- If I **include** updated.status (**with or without updated.tool**), I **always get a complain about missing updated.tool**.
- If I **exclude** updated.status but **include updated.tool**, I instead get a complaint that **updated.status is missing**.

As seen below, tool is not added into the patches array (and this cause the resolver to cry and give me the warning, and not proceeding with the bacjkend sync...)
![p](https://user-images.githubusercontent.com/63409162/107620815-98697480-6c55-11eb-9b0a-39624768053b.JPG)

Note: I tried to change the updated.tool value into something new (in case it was a thing with tool value not being changed, and consequently not being needed to be patched), but that did not work either. 

I understand that the patches array is constructed in the "immer" module, but I assume the updated.tool value is excluded before we get there... but I cannot seem to find WHERE in the code this happen *sigh*.

Any pointers would be much appreciated. I am really stuck atm and cannot release my next version to the client.

**Comment by iartemiev** [comment]
Hey @meducati, we've merged a PR that fixes this bug and the fix is available in `aws-amplify@4.0.2-unstable.5+3abfb8fc6`. It will be included in our next release.

**Comment by iartemiev** [comment]
The fix has been in aws-amplify@4.0.3. Closing the issue.

**Comment by meducati** [comment]
Perfect. Thank you! Will check it out next time I upgrade :)
